Settings: Enable ROM registration of handlers

Laczen opened the pull request:
Add the possibility to register static handlers (in ROM) using a new macro
`SETTINGS_STATIC_HANDLER_DEFINE(hname, name, get, set, commit, export)` an example for settings stored under `ps/data` is:
`SETTINGS_STATIC_HANDLER_DEFINE(ps_data_handler, "ps/data", NULL, ps_set, ps_commit, NULL)`

To maintain support for dynamic handlers
`CONFIG_SETTINGS_RAM_HANDLERS`must be enabled, which is by default.

When using static handlers there is no check if this handler has
been registered earlier, the latest registered static handler will be
considered valid for any set/get routine, while the commit and export
routines will be executed for both registered handlers.

When a dynamic handler is registered a check is done to see if there was
an earlier registration of the name in all static and dynamic handlers, if there was the
registration will fail.

To get to the lowest possible RAM usage it is advised to set
`CONFIG_SETTINGS_DYNAMIC_HANDLERS=n`.

Laczen committed:
subsys/settings: Enable handler ROM registration

Add the possibility to register handles to ROM using a new macro
SETTINGS_REGISTER_STATIC(handler), the handler is of type
settings_handler_stat and has to be declared as const:

```
const struct settings_handler_stat test_handler = {
	.name = "test", /* this can also be "ps/data"
	.h_get = get,
	.h_set = set,
	.h_commit = NULL, /* NULL defines can be ommited */
	.h_export = NULL  /* NULL defines can be ommited */
};
SETTINGS_REGISTER_STATIC(test_handler);
```

To maintain support for handlers stored in RAM (dynamic handlers)
`CONFIG_SETTINGS_DYNAMIC_HANDLERS`must be enabled, which is by default.

When registering static handlers there is no check if this handler has
been registered earlier, the latest registered static handler will be
considered valid for any set/get routine, while the commit and export
routines will be executed for both registered handlers.

When a dynamic handler is registered a check is done to see if there was
an earlier registration of the name as a static or dynamic handler
registration will fail.

To get to the lowest possible RAM usage it is advised to set
`CONFIG_SETTINGS_DYNAMIC_HANDLERS=n`.

Updates:

a. Changed usage of RAM to DYNAMIC/dynamic, ROM to STATIC/static
b. Updated settings.h to remove added #if defined()
c. Make static handlers always enabled
d. Corrected error introduced in common-rom.ld.
e. Changed return value of settings_parse_and_lookup to
   settings_handler_stat type to reduce stack usage.
f. Updated the name generated to store a handler item in ROM. It now
   uses the name used to register in combination with the line where
   SETTINGS_REGISTER_STATIC() is called.
g. renamed settings_handler_stat type to settings_handler_static
h. renamed SETTINGS_REGISTER_STATIC to SETTINGS_STATIC_HANDLER_DEFINE()

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Vudentz reviewed (changes requested):
We can do even better and declare the variable within the macro, see how Ive done that in GATT:

https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/gatt.h#L432

Also the kernel usually refer to static vs dynamic declaration so we might as well stick to that terminology instead of using RAM/ROM.

Vudentz commented on subsys/settings/Kconfig:
```diff
@@ -24,6 +24,20 @@ config SETTINGS_RUNTIME
 	help
 	  Enables runtime storage back-end.
 
+config SETTINGS_ROM_HANDLERS
```
I expected that static ones are always available to use otherwise we might have to make macros to fail at build time if the config setting is not selected which imo has no real benefit.

Also we these 2 options one can actually enable the settings without enabling either types of handles which would turn the settings useless since no handlers can be registered.

Laczen replied:
I tried to get the code "compatible" with the present case. If you now set SETTINGS_ROM_HANDLERS to n you should get the exact same size as we get in the present code. On the otherside to generate the smallest RAM usage you can set SETTINGS_RAM_HANDLERS to n.

When they are both set to n you will not be able to call SETTINGS_ROM_REGISTER() or settings_register() so compilation would fail.

I agree that it might lead to confusion, if we can agree on always enabling the ROM (static) handling, I will remove the option SETTINGS_ROM_HANDLERS.

Vudentz replied:
Id go for having static support allways when CONFIG_SETTINGS is set, this makes it a lot simpler in terms of config options and I assume the vast majority of the handlers can be tranformed in static handlers quite easily, to be honest I think the dynamic handlers might only be useful for application or to override but you said that is not even possible right now.

Laczen replied:
OK, I will by default enable the static handlers.

Vudentz commented on include/linker/common-rom.ld:
```diff
@@ -100,8 +100,17 @@
+
 #if defined(CONFIG_BT_SETTINGS)
-	SECTION_DATA_PROLOGUE(_bt_settings_area,,SUBALIGN(4))
+	SECTION_DATA_PROLOGUE(_settings_area,,SUBALIGN(4))
```
This seems wrong, the bt_settings_area has nothing to do with the settings handlers, it has a different type bt_settings_handler.

If the idea is to have just one handler and reuse on bluetooth side it should not be conditional to CONFIG_BT_SETTING, instead it should probably be CONFIG_SETTINGS and then remove the bt_settings_handler altogether.

Laczen replied:
Yes, this is wrong, it entered during development and I have forgotten to correct it back to `_bt_settings_area`. Thanks for catching this.

In the long run, I would expect the handler registration for bt_settings to disappear, I would keep the rework to one settings_handler region for a future PR.

Vudentz commented on include/settings/settings.h:
```diff
@@ -42,8 +42,11 @@ typedef ssize_t (*settings_read_cb)(void *cb_arg, void *data, size_t len);
  * These are registered using a call to @ref settings_register.
  */
 struct settings_handler {
```
Id add a settings_handler_static variant without the node field which is what should be used by the macro when statically defining things so we don't take more space than needed.

Laczen replied:
This could be done, but would have a big influence on the code, now throughout the code there is only one settings_handler type that is shared for static and dynamic handlers and in the case of "bestmatch" comparison this is shared by both static and dynamic handlers. Actually if we split up the settings_handler type we might even be using more ROM (code size) handling two different settings_handler structs than we are saving.

Vudentz commented on include/settings/settings.h:
```diff
@@ -93,6 +96,18 @@ struct settings_handler {
 	 */
 };
 
+#if defined(CONFIG_SETTINGS_ROM_HANDLERS)
```
Afaik #if defined are not recommended on header since that means the documentation should also be conditional.

Laczen replied:
I think I can safely remove all of them in settings.h, with the disadvantage of using a little more space for the static handlers.

Vudentz commented on include/settings/settings.h:
```diff
@@ -104,15 +119,16 @@ struct settings_handler {
  */
 int settings_subsys_init(void);
 
+#if defined(CONFIG_SETTINGS_RAM_HANDLERS) /*  */
```
ditto, for functions it should be fine to always have it available in the header since the build will fail when linking.

Laczen replied:
OK, will change.

carlescufi replied:
And the doc will also fail (like it's failing now) otherwise.

Vudentz commented on subsys/settings/src/settings.c:
```diff
@@ -17,25 +17,40 @@
 
 	k_mutex_lock(&settings_lock, K_FOREVER);
 
+#if defined(CONFIG_SETTINGS_ROM_HANDLERS)
```
You should probably document that static defined handlers takes precedence over dynamic ones, though this eliminates the possibility of overriding system handlers, either way Im fine we just have to clearly state what are their priority somewhere.

Laczen replied:
Well, it is the other way around, dynamic handlers take precedence over static ones, if it finds a match after the ROM handlers in the RAM handlers it will take the RAM handlers. I had the idea of overriding handlers in mind when writing.

Vudentz replied:
The overriding is something that can be used by mistake though so perhaps it is fine to not allow it at runtime, for the static one I guess we cannot detect duplicate handlers though expect perhaps if we declare some variable with the handler prefix to cause build time errors but that will increase the code size so I don't think it is worth it.

Vudentz commented on subsys/settings/src/settings_store.c:
```diff
@@ -110,6 +109,19 @@ int settings_save(void)
+	}
+#endif /* CONFIG_SETTINGS_ROM_HANDLERS */
+
+#if defined(CONFIG_SETTINGS_RAM_HANDLERS)
```
Ditto, btw you might as well create a helper iterator, e.g. settings_foreach_handler, to make sure that everytime it is iterating in the same order.

Laczen commented:
> We can do even better and declare the variable within the macro, see how Ive done that in GATT:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/gatt.h#L432
>

I was trying to get the api for ROM handlers to closely match the RAM part. What I was thinking to make it completely the same is to register a pointer to the handler in ROM instead of the handler, however my understanding of the `Z_STRUCT_ITERABLE` is very limited.

> Also the kernel usually refer to static vs dynamic declaration so we might as well stick to that terminology instead of using RAM/ROM.

Yes, I will change all RAM into dynamic and ROM into static.

carlescufi commented:
Ooops, @Laczen seems we deleted your branch in this repo and this caused this PR to be closed, sorry about that! Can you please create a new one from your fork?

nashif commented:
I am not sure what is going on here, this PR is against an unofficial branch on the zephyr main tree on GH, why are you doing that? All your branches need to be on a fork, not on the main tree. Please recreate this PR on a fork. 

I did not notice this dependency chain, the PR was closed automatically by GH once the invalid branch was removed.

carlescufi commented:
> I am not sure what is going on here, this PR is against an unofficial branch on the zephyr main tree on GH, why are you doing that? All your branches need to be on a fork, not on the main tree. Please recreate this PR on a fork.
> 
> I did not notice this dependency chain, the PR was closed automatically by GH once the invalid branch was removed.

I fixed it now by retargeting the PR against master. That said, the first 2 commits need to be deleted from the forked branch.

Laczen commented:
> I am not sure what is going on here, this PR is against an unofficial branch on the zephyr main tree on GH, why are you doing that? All your branches need to be on a fork, not on the main tree. Please recreate this PR on a fork.
> 
> I did not notice this dependency chain, the PR was closed automatically by GH once the invalid branch was removed.

Sorry for this, I didn't even know that I did this.

Laczen commented:
> We can do even better and declare the variable within the macro, see how Ive done that in GATT:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/gatt.h#L432

@Vudentz, I was afraid that in case of a static handler it would occupy two regions in ROM, once as a const and once in the handlers list, however after studying the rom report I can only find it in the handlers list. So I think the proposed solution is good and I don't know if it's possible to do it better (unless I completely misunderstood what you meant).

Vudentz commented:
> > We can do even better and declare the variable within the macro, see how Ive done that in GATT:
> > https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/gatt.h#L432
> 
> @Vudentz, I was afraid that in case of a static handler it would occupy two regions in ROM, once as a const and once in the handlers list, however after studying the rom report I can only find it in the handlers list. So I think the proposed solution is good and I don't know if it's possible to do it better (unless I completely misunderstood what you meant).

If you take the structs we use in GATT:

/** @brief GATT Service structure */
struct bt_gatt_service_static {
	/** Service Attributes */
	const struct bt_gatt_attr *attrs;
	/** Service Attribute count */
	size_t			attr_count;
};

/** @brief GATT Service structure */
struct bt_gatt_service {
	/** Service Attributes */
	struct bt_gatt_attr	*attrs;
	/** Service Attribute count */
	size_t			attr_count;
	sys_snode_t		node;
};

We use the first one as the in_section format so we save 4 bytes per service, with that one can just cast to the later if we have to, though there are no casts since we have dedicated code for iterating these in bt_gatt_foreach_attr_type.

That said maybe if we don't have too many handlers then saving 4 bytes of the node might not make sense, actually it may not save 4 bytes depending if your struct will require paddings for aligment.

Laczen commented:
> > > We can do even better and declare the variable within the macro, see how Ive done that in GATT:
> > > https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/gatt.h#L432
> > 
> > 
> > @Vudentz, I was afraid that in case of a static handler it would occupy two regions in ROM, once as a const and once in the handlers list, however after studying the rom report I can only find it in the handlers list. So I think the proposed solution is good and I don't know if it's possible to do it better (unless I completely misunderstood what you meant).
> 
> If you take the structs we use in GATT:
> 
> /** @brief GATT Service structure _/ struct bt_gatt_service_static { /_* Service Attributes _/ const struct bt_gatt_attr _attrs; /__ Service Attribute count */
> size_t	attr_count;
> };
> 
> /** @brief GATT Service structure _/ struct bt_gatt_service { /_* Service Attributes _/ struct bt_gatt_attr	_attrs; /__ Service Attribute count */
> size_t	attr_count;
> sys_snode_t	node;
> };
> 
> We use the first one as the in_section format so we save 4 bytes per service, with that one can just cast to the later if we have to, though there are no casts since we have dedicated code for iterating these in bt_gatt_foreach_attr_type.
> 
> That said maybe if we don't have too many handlers then saving 4 bytes of the node might not make sense, actually it may not save 4 bytes depending if your struct will require paddings for aligment.

Thanks for elaborating, I have updated the PR to include this proposal. At the same time I have changed the `settings_parse_and_lookup` to return the handler without the node information. The only thing left is to create a helper iterator but I would keep this for future improvement.

Laczen committed:
subsys/settings: Update bluetooth module

Updated the bluetooth module to use static handlers. Removed the
old bt specific static registration.

The routine bt_settings_init() is still calling settings_init() which
IMO is not needed anymore.

Updates:

changed SETTINGS_REGISTER_STATIC() to SETTINGS_STATIC_HANDLER_DEFINE()
changed settings_handler_stat type to settings_handler_static type
removed NULL declarations
renamed bt_handler to bt_settingshandler, as bt_handler already exists.
renamed all bt_XXX_handler to bt_xxx_settingshandler to avoid any
overlap.
changed SETTINGS_STATIC_HANDLER_DEFINE() to create variable names from
_hname by just prepending them with settings_handler_.
updated all bt_xxx_settings_handler to just bt_xxx.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

jhedberg commented on include/settings/settings.h:
```diff
@@ -93,6 +149,19 @@ struct settings_handler {
+ *	  _handler and the linenumber of SETTINGS_REGISTER_STATIC()
+ *
+ */
+#define SETTINGS_REGISTER_STATIC(_handler)	\
```
There are quite many similar "registration" concepts in the tree. It'd be good to use consistent naming throughout the project, and I don't think anything is currently using "register" as the verb, rather "define": `LOG_BACKEND_DEFINE` `BT_GATT_SERVICE_DEFINE` `K_THREAD_DEFINE` (and the `BT_SETTINGS_DEFINE` which you're removing in the subsequent patch). Most kernel objects also have `*_DEFINE()` macros, although they're not exactly registering anything (beyond kernel object debugging which will be able to iterate these objects later if necessary). So I'm wondering is there a good reason to deviate from this naming convention for settings handlers?

Laczen replied:
Agreed, I will change this to SETTINGS_STATIC_HANDLER_DEFINE(), I am unsure if I should change `settings_register` to `settings_dynamic_handler_define()`.

Vudentz replied:
Ive also suggested to just pass the handlers parameters and then do the handler declaration within the macro, like most DEFINE macro do, since most of the time the code using it is not doing anything with the variable.

Laczen replied:
I tried to keep it more inline with the `settings_register()` routine. Most defines were also from the pre `Z_STRUCT_SECTION_FOREACH()` definition, with this it seems more clear to define the struct and than to define using the struct. It also allows the removal of any NULL declaration.
If you would rather like: `SETTINGS_STATIC_HANDLER_DEFINE(bt_ccc, "bt/ccc", NULL, bt_ccc_set, NULL, NULL)` I can change it, no problem.

However I cannot stringify "bt/ccc" so it will need to be added, as well as bt_ccc.

Laczen replied:
I have to say there is another advantage to using `SETTINGS_STATIC_HANDLER_DEFINE(bt_ccc, "bt/ccc", ...)`: shippable failed on my PR because I defined `const struct settings_handler_static bt_handler={...`, `bt_handler` was already defined. This would not have been the case when I would have used your suggestion.

Laczen replied:
Reworked to `SETTINGS_STATIC_HANDLER_DEFINE(hname, name, get, set, commit, export)` because `native_posix` didn't seem to like the `const struct settings_handler_static` method.

jhedberg commented on include/settings/settings.h:
```diff
@@ -42,8 +42,64 @@ typedef ssize_t (*settings_read_cb)(void *cb_arg, void *data, size_t len);
+ * Config handlers without the node element, used for static handlers.
+ * These are registered using a call to SETTINGS_REGISTER_STATIC().
+ */
+struct settings_handler_stat {
```
I think it'd be better to be consistent with either "static" everywhere or then "stat" everywhere, but not mix them. I'm in favour of the former like you've done e.g. with SETTINGS_REGISTER_STATIC

Laczen replied:
I will try to get rid of the `settings_handler_stat` type by making it the settings_handler type and to create a settings_handler_dynamic as the combination of a static handler and a node element.

Laczen replied:
Done

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -45,6 +45,7 @@
 #include "settings.h"
 #include "gatt_internal.h"
 
+
```
This line addition seems unrelated? (and redundant)

Laczen replied:
Yes, I will remove this line.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3606,7 +3607,15 @@ static int ccc_set(const char *name, size_t len_rd, settings_read_cb read_cb,
-BT_SETTINGS_DEFINE(ccc, ccc_set, NULL, NULL);
+const struct settings_handler_stat bt_ccc_handler = {
+	.name = "bt/ccc",
+	.h_get = NULL,
```
Any struct members not explicitly initialised are implicitly initialised to NULL. At least in the Bluetooth code we don't generally list any members in the initialisation that we don't use. I don't have a super strong opinion on this though, but it'd be good to keep some consistency throughout the code base.

Laczen replied:
As you wish, I added them (consistently) to the bluetooth module to allow developers to quickly see how the elements are called. Since in the bluetooth code they are generally omitted I will do the same.

jhedberg reviewed (changes requested):
Some cosmetic, mostly naming related comments

Vudentz commented on subsys/settings/src/settings_store.c:
```diff
@@ -110,6 +109,17 @@ int settings_save(void)
+	}
+
+#if defined(CONFIG_SETTINGS_DYNAMIC_HANDLERS)
+	struct settings_handler *ch;
```
You may as well put this block into a helper function e.g setting_save_dynamic to avoid having declarations in the middle of the scope, though the later may just be a convention no really enforced so I fine you decide it doesn't justify writing a new function.

Laczen replied:
I don't think it justifies writing a new function. I have put it there to remove any warning about unused variables if CONFIG_SETTINGS_DYNAMIC_HANDLERS = n. I could of course have added the #if twice but I think it's easier to read this way.

Laczen commented:
@Vudentz, @jhedberg, if you are happy with the applied change requests could you approve the PR. If you would like someone else to have a look before merging I will add the DNM property.

Vudentz reviewed (approved):
LGTM

Vudentz commented on include/settings/settings.h:
```diff
@@ -93,6 +149,28 @@ struct settings_handler {
 
+/**
+ * Define a static handler for settings items
+ *
```
Just a minor addition here, you might want to add that us the variable name the macros expands to just in case the code wants to make use of it for some reason/

Laczen replied:
@Vudentz, thanks. I don't quite understand your last addition. Can I return the created variable name ? And how would it be used (I am not asking about the why it would be used) ? My knowledge of c (and macro's) is still limited but I'm always open to learn.

nvlsianpu commented on include/settings/settings.h:
```diff
@@ -42,8 +42,64 @@ typedef ssize_t (*settings_read_cb)(void *cb_arg, void *data, size_t len);
  */
 struct settings_handler {
+
+	char *name;
```
I would extract common structure part for static and dynamic handler (setting_handler_base/common) and then declare static and structure based on that.

function like settings_parse_and_lookup() might even use it as return value type.

Laczen replied:
I tried doing something like that. The problem is that the node info is part of the dynamic handler, so you have to define the dynamic handler as `struct settings_handler test = {...}`
if you would make the dynamic handler as the combination of static handler + node, you would need to define the above as `struct settings_handler test = { .static.get = ...}` this would just seem strange, and it would be an API change for dynamic handlers.
I found the way it is now the simplest and cleanest.
BTW: I have changed the return type of `settings_parse_and_lookup()` to the static type.

nvlsianpu replied:
OK, I can live with how this is implemented.

nvlsianpu replied:
Although both struct have to be almost the same - If somebody changes only one of them, then will introduce a bug.

nvlsianpu replied:
> this would just seem strange, and it would be an API change for dynamic handlers.

Actually was thinking about making it symmetric.
```C
struct settings_handler_basic {
    char *name;
   ...
};

struct settings_handler_static {
    struct struct settings_handler_basic base;
};

struct settings_handler {
    struct settings_handler_basic base;
    sys_snode_t node;
    /**< Linked list node info for module internal usage. */
};

```

nvlsianpu replied:
I men I can accept API change. This is not critical so can resign from this, just let me know whether you think that no wort to do so.

Laczen replied:
It would still be strange to create a dynamic settings_handler:
```
static struct settings_handler ps_handler = {
  .base.name = "ps",
  .base.h_get = get,
};
```
As the settings module is already complicated to use I would not do this. I think the next thing to do is to remove the dynamic handlers and only to support static ones.

nvlsianpu reviewed (commented):
I reviewed the settings submodule patch part. The rest will be reviewed on Wednesday.

Laczen commented:
@Vudentz, I have simplified the name generation for the variable, just put settings_handler_ before the name. This is simple and gives direct access to the variable name.

carlescufi commented:
@jhedberg @nvlsianpu can you take another look please?

nvlsianpu reviewed (approved):
@Laczen Your decision is whether you want to apply my suggestion:
https://github.com/zephyrproject-rtos/zephyr/pull/16897/files#r297673926

I approved this as
If not - patch can be merged right now.
